### DISCUSSION TITLE
Call this plugin API from another plugin

### DISCUSSION THREAD

**Discussion by LeonardoGentile** [Taitava/obsidian-shellcommands#396] [Q&A]
Hello, 
I'm new to obsidian plugin development.  
I'd like to create a plugin that calls a specific app from the command line and after elaborating its return it passes it back to obsidian. 

Given that I'll have to interact with the OS via bash/terminal I though it would be a good idea not to reinvent the wheel and use this plugin as a requirement for the new one. 

1 Is there a way to call this plugin API from another plugin?
2 Given I'll use a minimum of features of this plugin would it an overkill?
3 What are the best practices in general to make plugins communicate with each other?

Thanks 🙏

**Comment by Taitava** [comment]
Hi! Thank you for your interest about calling this plugin from another plugin via an API. 🙂 

> 1 Is there a way to call this plugin API from another plugin?

Currently, this plugin does not have an API that other plugins could call. I am interested in hearing ideas of what features an API could offer. Maybe I could then begin planning a first sub-set of API features. Anyway, it could take quite a long time before a conrete feature could come out, as I currently do not have a clear picture of the first steps towards building an API.

> 2 Given I'll use a minimum of features of this plugin would it an overkill?

It might be an overkill, but it depends very much on what you want to do. The decision can be evaluated e.g. by the following questions:

**What features of the SC plugin would you like to utilize via an API from your plugin?**
  - a) Execute a shell command and utilize any of the [output handlers](https://publish.obsidian.md/shellcommands/Output+handling/Output+channels#All+output+channels)? (You mentioned you'd like to pass the output back to Obsidian).
  - b) Execute a shell command when an [event is triggered](https://publish.obsidian.md/shellcommands/Events/Events+-+general+principles)?
  - c) Utilize the [`{{variable}}` system](https://publish.obsidian.md/shellcommands/Variables/All+variables)?
  - d) Ask for input values via [Prompts](https://publish.obsidian.md/shellcommands/Preactions/Prompts)?

**If you don't need many of the above features, will it be sufficient for you to use [Node.js's `child_process` functions](https://nodejs.org/api/child_process.html) directly?**
- The SC plugin uses [`child_process.spawn()`](https://nodejs.org/api/child_process.html#child_processspawncommand-args-options), but maybe you could get started with [`child_process.exec()`](https://nodejs.org/api/child_process.html#child_processexeccommand-options-callback), which I think is a bit simpler than `spawn()`.
- In case you are willing to see how the SC plugin calls `spawn()` to execute shell commands, you can explore the source code, starting from [`Shell.spawnChildProcess`](https://github.com/Taitava/obsidian-shellcommands/blob/8d030a23540d587a85bd0dfe2e08c8e6b6b955ab/src/shells/Shell.ts#L177-L180), and continuing to `ShellCommandExecutor`.

> 3 What are the best practices in general to make plugins communicate with each other?

I'm not familiar with the best practises in this regard, and if I will ever create an API, I would first need to research this topic. If I will end up planning an API, I could take a look at the [Dataview](https://github.com/blacksmithgu/obsidian-dataview) plugin's API and see how it's implemented - just to name one example. There can be other plugins with valuable information, too.
